src/NuGet.Core/NuGet.Protocol/Utility/LocalFolderUtility.cs

```diff
@@ -1298,5 +1269,33 @@ public static void GenerateNupkgMetadataFile(string nupkgPath, string installPat
                     }
                 });
         }
+
+        private static IEnumerable<T> CancellableYieldEnumaration<T>(IEnumerator<T> enumerator, ILogger log, CancellationToken cancellationToken)
```

Comment:
This is a clever idea to reduce the impact down stream, a few things, something like this is probably simpler:

``` C#
    cancellationToken.ThrowIfCancellationRequested();

    try
    {
        foreach (var item in enumerator)
        {
            yield return item;

            cancellationToken.ThrowIfCancellationRequested();
        }
    }
    catch (Exception ex) when (ex is not OperationCanceledException)
    {
        log.LogWarning(e.Message);
    }
```

Otherwise,  you are missing a dispose on the IEnumerator<T> (which forever does for you) - it holds a handle that needs to be released.

Reply:
One problem here is we can't yield from try catch [CS1626](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1626?f1url=%3FappId%3Droslyn%26k%3Dk(CS1626)), so I ended up current code.  I refactored code to do `IEnumerator` dispose with `using` statement, please check.

Actually, we can do something simpler. What do you think?

```
        private static T[] CancellableEnumaration<T>(IEnumerable<T> enumerable, ILogger log, CancellationToken cancellationToken)
        {
            try
            {
                var result = new List<T>();
                foreach (var item in enumerable)
                {
                    result.Add(item);
                    cancellationToken.ThrowIfCancellationRequested();
                }

                // .ToArray is necessary because previously it returns Array
                return result.ToArray();
            }
            catch (Exception e) when (e is not OperationCanceledException)
            {
                // On cancellation we bubble up exception to call stack.
                // Otherwise return all or nothing. If no exception return all.
                // Break on first exception with logging in order to keep previous experience, return empty array, but don't throw.
                log.LogWarning(e.Message);
            }

            return Array.Empty<T>();
        }
```

Reply:
Another approach would be to just move the catch into a separate method:

``` C#
static IEnumerable<T> CancellableYieldEnumaration<T>(IEnumerable<T> enumerable, ILogger log, CancellationToken cancellationToken)
{
    cancellationToken.ThrowIfCancellationRequested();

    IEnumerator<T> enumerator = enumerable.GetEnumerator();

    try
    {
        while (MoveNext(enumerator, log))
        {
            yield return enumerator.Current;

            cancellationToken.ThrowIfCancellationRequested();
        }
    }
    finally
    {
        enumerator.Dispose();
    }
}

static bool MoveNext<T>(IEnumerator<T> enumerator, ILogger log)
{
    try
    {
        return enumerator.MoveNext();
    }
    catch (Exception e)
    {
        log.LogWarning(e.Message);
    }

    return false;
}
```

Reply:
Or a combo of your original change with mine - you need to call Dispose on the enumerator.

Reply:
Did you check my latest push?https://github.com/NuGet/NuGet.Client/blob/dab426f7a84218a4935c8d86dc8777ecfb231c3e/src/NuGet.Core/NuGet.Protocol/Utility/LocalFolderUtility.cs#L1283-L1330

In line 1304 `using (var enumerator = enumerable.GetEnumerator()) ` I believe it does dispose on the enumerator later. right?